modules/basegl/include/modules/basegl/processors/layerprocessing/layerglprocessor.h

```diff
@@ -0,0 +1,102 @@
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#pragma once
+
+#include <modules/basegl/baseglmoduledefine.h>
+
+#include <inviwo/core/processors/processor.h>
+#include <inviwo/core/ports/layerport.h>
+#include <modules/opengl/shader/shader.h>
+#include <modules/opengl/buffer/framebufferobject.h>
+
+namespace inviwo {
+
+class Layer;
+class DataFormatBase;
+class ShaderResource;
+class TextureUnitContainer;
+
+/*! \class ImageGLProcessor
+ *
+ * \brief Base class for layer processing on the GPU using OpenGL.
+ *
+ * The LayerGLProcessor provides the basic structure for layer processing on the GPU.
+ * Derived shaders have to provide a custom fragment shader which is used during rendering.
+ * Optionally, derived classes can overwrite LayerGLProcessor::postProcess() to perform
+ * post-processing of the layer data set in the outport. Furthermore, it is possible to
+ * be notified of changes in the input layer by overwriting LayerGLProcessor::afterInportChanged().
+ *
+ * \see ImageGLProcessor
+ */
+class IVW_MODULE_BASEGL_API LayerGLProcessor : public Processor {
+public:
+    LayerGLProcessor(std::shared_ptr<const ShaderResource> fragmentShader,
+                     Shader::Build build = Shader::Build::Yes);
+
+    virtual void initializeResources() override;
+
+    virtual void process() override;
+
+protected:
+    void markInvalid();
+
+    /*! \brief this function gets called right before the actual processing but
+     *         after the shader has been activated
+     *
+     * overwrite this function in the derived class to perform things like custom shader setup
+     */
+    virtual void preProcess(TextureUnitContainer& cont);
+
+    /*! \brief this function gets called at the end of the process function
+     *
+     * overwrite this function in the derived class to perform post-processing
+     */
+    virtual void postProcess();
+
+    /*! \brief this function gets called whenever the inport changes
+     *
+     * overwrite this function in the derived class to be notified of inport onChange events
+     */
+    virtual void afterInportChanged();
```

Comment:
Do we need to expose this?

Reply:
Yes, the derived processors may have properties depending on the input data.

https://github.com/inviwo/inviwo/blob/c696c112bf966e14dd90001cc70dcb1ba7974e02/modules/basegl/src/processors/imageprocessing/imagemapping.cpp#L88-L105
https://github.com/inviwo/inviwo/blob/c696c112bf966e14dd90001cc70dcb1ba7974e02/modules/basegl/src/processors/volumeprocessing/volumegradientmagnitude.cpp#L86-L100

Reply:
Ok. although the first one could just as well run in process from what I see...
For the second case I would rather just have all 4 channels always a give an error if the data does not have the channel specified, but sure.  We should generally try to avoid having properties depend on data as much as possible.